### PR TITLE
Fix resource leaking due to undeleted ngraph::Executable

### PR DESCRIPTION
**Pull request by XWangIntel**
There must be some kind of cyclic reference for shared_ptr\<ngraph::Executable\> and I'm still looking for the method to break it.
Before that the resource leaking can be fixed by replacing shared_ptr\<ngraph::Executable\> which raw pointer. It is not just a workaround but also reduce the overhead of incrementing shared_ptr use count, which can reach several hundreds and totally unnecessary. Therefore should be good to have anyway.

### CODE REVIEW

File: ngraph_bridge/ngraph_freshness_tracker.cc [hunk 1]
```diff
@@ -69,10 +69,10 @@ void NGraphFreshnessTracker::RemoveTensor(const void* base_pointer) {
 }
 
 void NGraphFreshnessTracker::RemoveUser(
-    std::shared_ptr<ngraph::runtime::Executable> user) {
+    const std::shared_ptr<ngraph::runtime::Executable>& user) {
   mutex_lock l(mu_);
-  for (auto kv : freshness_map_) {
-    kv.second.erase(user);
+  for (auto& kv : freshness_map_) {
```
Comment (XWangIntel) [inline ngraph_bridge/ngraph_freshness_tracker.cc]: change `kv` to `& kv` is the fix for this problem. The rest of change is just for performance and simplicity.
In short, if using `kv`, we just erase user from the copy of `ngraph::Executable` set. The original freshness_map is untouched. Period.

**Comment by XWangIntel** [comment]
@googlebot I signed it!

File: ngraph_bridge/ngraph_freshness_tracker.cc (hunk 1 above)
Comment (shresthamalik) [reply ngraph_bridge/ngraph_freshness_tracker.cc]: Good Catch! :)

File: ngraph_bridge/ngraph_freshness_tracker.h [hunk 2]
```diff
@@ -114,15 +114,15 @@ class NGraphFreshnessTracker : public ResourceBase {
...
+  void RemoveUser(const std::shared_ptr<ngraph::runtime::Executable>& user);
 
  private:
   // mutex protecting the freshness_map_
   mutex mu_;
   // for each base pointer (of tensor), maintains a set of ng::Functions that
   // use it
   // Each ng function in the set is then a user of the base_pointer
-  std::map<const void*, std::set<std::shared_ptr<ngraph::runtime::Executable>>>
+  std::map<const void*, std::set<ngraph::runtime::Executable*>>
```
Comment (avijit-nervana) [inline ngraph_bridge/ngraph_freshness_tracker.h]: We don't want to save raw pointers - as they may become null under the hood. We need to save a unique_ptr or shared_ptr. But in this case we are getting a shared_ptr from nGraph - so we are going to stay with shared_ptr.

File: ngraph_bridge/ngraph_freshness_tracker.cc (hunk 1 above)
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_freshness_tracker.cc]: Good catch! Thanks for debugging this :)

File: ngraph_bridge/ngraph_freshness_tracker.h (hunk 2 above)
Comment (XWangIntel) [reply ngraph_bridge/ngraph_freshness_tracker.h]: @avijit-nervana My concern is about performance. I see the use count of ngraph::Executable goes up to hundreds. And the increment/decrement is implemented with atomic operation and therefore the overhead is quit significant.
If we are sure that the lifespan of executable is longer than freshness tracker, it is still worth the change. You are definitely more familiar with that and can make the decision.

Comment (avijit-nervana) [reply ngraph_bridge/ngraph_freshness_tracker.h]: @XWangIntel I agree with you on this. I personally do not like shared ptrs just for this reason. We will do this clean up a few weeks later - so will track this definitely.

**Comment by avijit-nervana** [comment]
There are compilation errors. See here: https://buildkite.com/ngraph/external-fork/builds/2#dab399bd-427a-4da4-a272-4d06d19707c2/6-6206

Please run build_ngtf.py and test_ngtf.py to ensure that it builds and sanity tests pass.

**Comment by XWangIntel** [comment]
@googlebot I consent

**Comment by XWangIntel** [comment]
> There are compilation errors. See here: https://buildkite.com/ngraph/external-fork/builds/2#dab399bd-427a-4da4-a272-4d06d19707c2/6-6206
> 
> Please run build_ngtf.py and test_ngtf.py to ensure that it builds and sanity tests pass.

Sorry this change is only verified on full stack build of NNP-transformer, and then moved to the new repo. Will verify that in standalone repo.

**Comment by XWangIntel** [comment]
@googlebot I consent

**Comment by XWangIntel** [comment]
@googlebot I consent!

**Comment by XWangIntel** [comment]
cla:yes

**Comment by XWangIntel** [comment]
@googlebot

**Comment by XWangIntel** [comment]
@googlebot I verified it!

**Comment by XWangIntel** [comment]
Submit the same PR from the repo instead of forked one.

**Pull request closed without merging**